### ISSUE TITLE
Editor touch controls

### ISSUE THREAD

**Issue by firelizzard18** [Microsoft/vscode#47770]
Issue Type: <b>Feature Request</b>

When using a touch screen, I want to be able to interact with VSCode in an intuitive manner.

When I touch and drag on text, I expect it to do something. I think the most intuitive option would be to select text.

When I touch and drag on empty space (space with no characters in the buffer), I expect the editor to scroll with my finger.

VS Code version: Code 1.22.1 (950b8b0d37a9b7061b6f0d291837ccc4015f5ecd, 2018-04-06T02:26:57.615Z)
OS version: Windows_NT x64 10.0.16299

**Comment by dumbdimb** [comment]
I was thinking of requesting this feature, too. I often use VS Code in my computer while remoting in from my phone using Remote Desktop Connection using "Touch" (not "Pointer" mode) and it would be really nice for the editor to behave like a native mobile app.

Given VS Code is open
And VS code knows I'm using touch input
When a context menu is opened
Then the options on the context menu are big enough to accommodate touch input

Given I have one of the left-side panes open
When I drag or swipe on the pane with one finger
Then the view should scroll with my finger

Given I have a left side pane open
When I tap on something 
Then the file or webpage relevant to what I tapped should be opened

Given a left side pane is open
When I press and hold something and then let go
Then the full context menu should show

Given I am on the file tree pane
When I press and hold on a file or folder then drag it across the pane
Then VS Code should let me move the file or folder to another directory in the tree

Given I have a file or webpage open in the editor/main view
When I drag or swipe on the view with one finger
Then the file or webpage should scroll with my finger

Given I have a file open in the editor
When I tap on some text
Then the text cursor should move to where I tapped
And a pin should show below the text cursor

Given the text cursor is visible somewhere in the editor window
When I drag the pin
Then the cursor should move along with the pin

Given the pin is visible
When I let go of the pin
Then there should be a small context menu above the text cursor for common tasks like "Select all", "Paste", and "..."

Given a small context menu is open
When I tap on "..."
Then the full context menu should show

Given I have a file open in the editor
When I double tap on some text
Then the word I tapped on should be highlighted 
And left and right pins should show
And a small context menu should show above my selection for common tasks like "cut", "copy", and "..." 

Given I have selected some text
And the left and right pins are showing
When I drag the pins
Then my selection should adjust to match where the selection pins are

Given I have a file or webpage open in the editor/main view
When I press and hold on the file or webpage and then let go
Then any selection I had should not clear
And I should see a full context menu

**Comment by kcambridge** [comment]
Please add this. I want to use VS Code but will have to stick with NP++  until this is added.

**Comment by vintproykt** [comment]
+
We definitely need this feature. I can't live without code) Getting in touch with new skills will be much easier when it's implemented!
By now you only can use "Expand selection" to select some word or scope, but it makes you stumbling just with the touchscreen, which sometimes is the only input device.

**Comment by vintproykt** [comment]
Plus there is pretty weird behavior with combo-boxes - like you can't even select anything (for example, choosing log target in console panel failed).

**Comment by tsmspace** [comment]
I'm really newbie, but I spent a good few minutes trying to figure out if touch-drag to select text was a possibility. I can see where the conflict between wiping the view up and down and dragging a cursor would be an issue. I wonder if there is a way to use two fingers instead?? I can touch with two fingers, then drag them around to move the two ends of the selection, and because zoom isn't really a feature in vscode (mostly if someone wants larger text they can change it that one time in options, normally you wouldn't be continously zooming and unzooming like some browser pages). 

for me the option of two fingers to select seems the most logical. 

I don't like the idea of touching, then holding, then selecting an option, then one at a time dragging each end back and forth, meanwhile accidentally de-selecting, meanwhile a trackpad is already across the finish-line,

**Comment by stockers1** [comment]
This wouls be a really great feature to have.  one of the most annoying issues i have with vscode.

**Comment by PetrKryslUCSD** [comment]
Is anything happening on this front?

**Comment by plezploz** [comment]
I need this too!

**Comment by gitappcomua** [comment]
I want it too!

**Comment by Yamilquery** [comment]
I need this feature too

**Comment by pedroni** [comment]
This issue happens while using other devices such as an iPad. I’m using https://github.com/cdr/code-server for developing on iPad and I’m facing this issue too.

**Comment by SashaOv** [comment]
Was trying VS Code on Surface Pro 8 tablet and came across this issue. Windows built-in Notepad does a decent job with touch support: you can edit text without keyboard by just using touch. VS Code cannot. Not being able to use VS Code for coding on tablet is very sad.

**Comment by PetrKryslUCSD** [comment]
Almost a year later, no progress. That is disappointing. Touch is really important on modern devices. I can get a lot done on my Surface Pro, but using VS Code by touch is a pain. Are there at least some plans?

**Comment by YamilDiaz** [comment]
We need to develop by ourselves.

**Comment by BenHunder** [comment]
Still looking for this!

**Comment by yiiman-dev** [comment]
Still looking for this too

**Comment by thefirstkfx** [comment]
Great feature to have when working on touch devices.

**Comment by udu3324** [comment]
Please add this feature!

**Comment by pedestriantactics** [comment]
Using codespaces with an iPad and an external keyboard and this is the one thing holding up my workflow. I can edit this post easier than I can edit my code right now 🫤

**Comment by cschoenecker** [comment]
This works best via Chrome on the iPad (although it is WebKit like Safari). The scroll works without setting the site zoom to 85% and the selection with the cursor also (the right click does not open the editing popup of the OS with cause the issue)

**Comment by jbschooley** [comment]
Yes, we really need this! I've been trying to find a way to get some of my dev work done on a tablet for weeks. On both my Windows tablet and my iPad Pro I can't select text by touching and dragging.

**Comment by cschoenecker** [comment]
since iOS 16 it is the display of the contextual menu (right click) that seems to crash the text selection on iPad. Concerning the scroll bug, it is fixed since iPas0S 16.4

**Comment by mauriciabad** [comment]
Given that there's no progress in this issue... Is there any extension that improves the touch capabilities of VS Code?

**Comment by cschoenecker** [comment]
While waiting for a resolution, it is better under Chrome (even with WebKit safari) but impossible in this case to have the PWA shortcut for the app in full screen

**Comment by ash-raph** [comment]
Can't wait to try it when it's ready

**Comment by MRoSlot** [comment]
Don't know if it useful but lady and gentleman please try unexpected keyboard app in android ctrl shift shortcut work really well

**Comment by ParadoxV5** [comment]
Text selection is possible. Steps:
1. Connect VSCode to WSL; it can be your current project or a new empty window
   * Doesn’t work without WSL connection
1. Tear your tab to a new window
   * Doesn’t work on the main window (where the side panels are)
1. Tap where you want to start the selection
   * Dragging doesn’t work
   * Tapping on empty space or clicking not only doesn’t work, but also disables scrolling when dragging on text in the next step.
1. Drag near where you want to end the selection
   * Be careful when dragging on empty space instead of text, because this will scroll the page in addition to selecting text.
   * From the perspective of scrolling:
     On tear-out windows, scrolling not only doesn’t work sometimes, but tends to instead select text consistently.

Very intuitive; thank you Microslop on the UX work.

* VS Code version: Code 1.134.0
* OS version: Windows_NT x64 10.0.26220
* Remote: WSL: Ubuntu
* Remote OS: Linux x64 6.6.87.2-microsoft-standard-WSL2

---

* Related: #81616
